cirq/experiments/xeb_theory.ipynb

```diff
@@ -0,0 +1,390 @@
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import cirq.google as cg\n",
+    "from cirq.experiments import random_quantum_circuit_generation as rqcg\n",
+    "\n",
+    "SQRT_ISWAP = cirq.ISWAP**0.5\n",
+    "\n",
+    "q0, q1 = cirq.LineQubit.range(2)\n",
+    "circuit = rqcg.random_rotations_between_two_qubit_circuit(\n",
+    "    q0, q1, \n",
+    "    depth=4, \n",
+    "    two_qubit_op_factory=lambda a, b, _: SQRT_ISWAP(a, b), \n",
+    "    single_qubit_gates=SINGLE_QUBIT_GATES\n",
+    ")\n",
+    "SVGCircuit(circuit)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Estimating fidelity\n",
+    "\n",
+    "Let $O_U$ be an observable that is diagonal in the computational\n",
+    "basis. Then the expectation of $O_U$ on $ρ_U$ is given by\n",
+    "\n",
+    "$$\n",
+    "    Tr(ρ_U O_U) = f ⟨𝜓_U|O_U|𝜓_U⟩ + (1 - f) Tr(O_U / D).\n",
+    "$$\n",
+    "\n",
+    "This equation shows how $f$ can be estimated, since $Tr(ρ_U O_U)$ can be\n",
+    "estimated from experimental data, and $⟨𝜓_U|O_U|𝜓_U⟩$ and $Tr(O_U / D)$ can be\n",
+    "computed.\n",
+    "\n",
+    "Let $e_U = ⟨𝜓_U|O_U|𝜓_U⟩$, $u_U = Tr(O_U / D)$, and $m_U$ denote the experimental\n",
+    "estimate of $Tr(ρ_U O_U)$. We can write the following linear equation (equivalent to the\n",
+    "expression above):\n",
+    "\n",
+    "$$\n",
+    "    m_U = f e_U + (1-f) u_U \\\\\n",
+    "    m_U - u_U = f (e_U - u_U)\n",
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Make long circuits (which we will truncate)\n",
+    "circuits = [\n",
+    "    rqcg.random_rotations_between_two_qubit_circuit(\n",
+    "        q0, q1, \n",
+    "        depth=100, \n",
+    "        two_qubit_op_factory=lambda a, b, _: SQRT_ISWAP(a, b), \n",
+    "        single_qubit_gates=SINGLE_QUBIT_GATES)\n",
+    "    for _ in range(10)\n",
+    "]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We will truncate to these lengths\n",
+    "cycle_depths = np.arange(3, 100, 9)\n",
+    "cycle_depths"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Execute circuits\n",
+    "Cross entropy benchmarking requires sampled bitstrings from the device being benchmarked *as well as* the true probabilities from a noiseless simulation. We find these quantities for all `(cycle_depth, circuit)` permutations."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pure_sim = cirq.Simulator()\n",
+    "noisy_sim = cirq.DensityMatrixSimulator(noise=cirq.depolarize(5e-3))\n",
+    "\n",
+    "records = []\n",
+    "for cycle_depth in cycle_depths:\n",
+    "    for circuit_i, circuit in enumerate(circuits):\n",
```

Comment:
I'm on the fence about this one. Using itertools does reduce the maximum line length for this cell, but only slightly since the itertools line is so long itself. I think the nested for loop may be more pedagogical. What do you think?

I'll re-write the `noisy_sim.sample(...` line anyways to be less long

Reply:
Yeah, that seems reasonable; keeping nested loops like that is easier to read.